proxy/src/telemetry/metrics/labels/mod.rs

```diff
@@ -127,6 +129,54 @@ pub struct DstLabels {
 #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
 pub struct TlsStatus(ctx::transport::TlsStatus);
 
+/// Labels describing a TLS handshake failure.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum HandshakeFailLabels {
+    Proxy {
+        /// Labels describing the TCP connection that closed.
+        transport: TransportLabels,
+        reason: HandshakeFailReason,
+    },
+    Control {
+        peer: Peer,
+        remote_addr: SocketAddr,
+        reason: HandshakeFailReason,
+    },
+
+}
+
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum HandshakeFailReason {
+    Io(Option<Errno>),
+    Tls(TlsError),
+}
+
+mk_err_enum! {
+    /// Translates `rustls::TLSError` variants into a `Copy` type suitable for
+    /// labels.
+    #[allow(non_camel_case_types)]
+    #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+    enum TlsError from &'a tls::Error {
+        tls::Error::InappropriateMessage {..} => INAPPROPRIATE_MESSAGE,
+        tls::Error::InappropriateHandshakeMessage {..} => INAPPROPRIATE_HANDSHAKE_MESSAGE,
+        tls::Error::CorruptMessage => CORRUPT_MESSAGE,
+        tls::Error::CorruptMessagePayload(_) => CORRUPT_MESSAGE_PAYLOAD,
+        tls::Error::NoCertificatesPresented => NO_CERTIFICATES_PRESENTED,
+        tls::Error::DecryptError => DECRYPT_ERROR,
+        tls::Error::PeerIncompatibleError(_) => PEER_INCOMPATIBLE,
+        tls::Error::PeerMisbehavedError(_) => PEER_MISBEHAVED,
+        tls::Error::AlertReceived(_) => ALERT_RECEIVED,
+        tls::Error::WebPKIError(_) => WEBPKI_ERROR,
```

Comment:
If the certificate validation failed then we should report what the certificate validation error was.

Reply:
e69c3fadbfe9534929029d0b7f1790fc580a5a23 adds variants for all the variants of `webpki::Error`. Is that sufficient?